### ISSUE TITLE
Promote `error` to custom type error

### ISSUE THREAD

**Issue by crockeea** [goldfirere/singletons#160]
With the addition of custom type errors, it should be possible to promote calls to `error` to a custom type error. For example, (very roughly)

```
singletons [d|
  foo :: Int -> Int
  foo 0 = 1
  foo _ = error $ "Can't call foo on " ++ (show x)
  |]
```

should get a type family

```
type family FFoo x where
  FFoo Z = S Z
  FFoo x = TypeError (Text "Can't call foo on " :<>: ShowType x)
```

It'd be great to change the `foo` in the error message to `FFoo`; any ideas how that can happen?

**Comment by goldfirere** [comment]
I had initially thought this was easy, but now I'm backing off.

It would be easy to take `error "blah"` and promote that to `TypeError (Text "blah")`, but now that you introduce function calls, etc., it's getting harder. This effectively requires that `String`s be promoted either to `Symbol`s or to `ErrorMessage`s, depending on the context. This would be hard.

I think the Right Way to do this is to promote `Char` to `Symbol` (where each such `Symbol` is one-character long) and then `String` to `[Symbol]`. Then we can concatenate type-level strings. `error` can be treated as `TypeError` composed with `Text`. Then it all should work. But this is not a quick fix, I'm afraid.

**Comment by RyanGlScott** [comment]
One other avenue we could explore is using `base-4.10.0.0`'s new [`AppendSymbol` type family](http://git.haskell.org/ghc.git/blob/69d9081d9fa3ff36fda36e4e11ef7e8f946ecf2a:/libraries/base/GHC/TypeLits.hs#l140) to append `Symbols` together.

**Comment by goldfirere** [comment]
But how do we know whether to promote `++` to `:++` or to `AppendSymbol`?

Really, the problem is that `Symbol` is *not* a promoted string. It is Something Else.

**Comment by RyanGlScott** [comment]
Perhaps we shouldn't be using `(++)` at all for appending `Symbol`s. They are abstract types, after all. I'm envisioning adding `Monoid` to `singletons` ([as you do in the `monads` branch](https://github.com/goldfirere/singletons/commit/d09004ef25590cb0748f67842c1549d1f3c10cab#diff-5ba76c7731f041b351bdcd590756fd1eR120)) and promoting/singling a `Monoid Symbol` instance that uses `AppendSymbol` as `mappend`. Then, you can promote this code:

```haskell
singletons [d|
  foo :: Int -> Int
  foo 0 = 1
  foo _ = error $ "Can't call foo on " <> (show x)
  |]
```

It's a bit different than how you'd probably write this unpromoted, but that's `Symbol` for you.

**Comment by goldfirere** [comment]
I like that plan. It still doesn't address the "custom type error" desire, but it's a step in the right direction.

**Comment by RyanGlScott** [comment]
It's worth noting that we have a promoted version of `Show` now in `singletons`, so implementing this becomes a lot easier. However, there's now a sort of ambiguity when it comes to promoting `error`. Should this:

```haskell
error $ "Can't call foo on " ++ (show x)
```

Be promoted to this:

```haskell
TypeError (Text "Can't call foo on " :<>: ShowType x)
```

Or this?

```haskell
Error ("Can't call foo on " :<> Show x)
```

Both are equally valid choices, but with different outcomes. The former choice gives you a proper type error, whereas the latter choice gives you a stuck type. Depending on what you're doing, you may want one or the other some of the time.

To support both use cases, I propose simply adding value-level counterparts for `TypeError` and `ErrorMessage`:

```haskell
typeError :: ErrorMessage -> a
text :: Symbol -> ErrorMessage
showType :: a -> ErrorMessage
(<>:) :: ErrorMessage -> ErrorMessage -> ErrorMessage
($$:) :: ErrorMessage -> ErrorMessage -> ErrorMessage
```

And grant the ability to promote them to their type-level counterparts.

**Comment by goldfirere** [comment]
How do we know whether to promote `++` to `:<>` or `:++`?

**Comment by RyanGlScott** [comment]
Ack, I mistyped. I should have written:

```haskell
error $ "Can't call foo on " <> (show x)
``` 

I'm certainly not proposing that we intelligently promote `(++)` to `(:<>)` or `(:++)` depending on the context. Let's do the obvious thing of promoting `(++)` to `(:++)`, `(<>)` to `(:<>)`, and `(<>:)` to `(:<>:)`.

**Comment by goldfirere** [comment]
So users wishing to promote a call to error will have to be mindful about using `<>` instead of `++` to concatenate their strings.

This plan now seems reasonable and implementable to me. The new symbols should be exported from either `Data.Singletons.TypeLits` or a new `Data.Singletons.TypeError`. Probably the former, given that `TypeError` itself is in `GHC.TypeLits`.

Thanks for pushing on this!

**Comment by RyanGlScott** [comment]
There's a wrinkle with this plan. Because of #29, `singletons` deliberately avoids promoting `($$:)` to `(:$$:)`. I'm not sure how to work around this...

**Comment by RyanGlScott** [comment]
Never mind, what I wrote in the previous comment isn't an issue. I'm going to hardwiring the promoted definitions anyways, and `type ($$:) = '(:$$:)` is a completely appropriate hack for this situation.

**Comment by RyanGlScott** [comment]
I'm going to give up on this for now, since I think this idea is blocked on #82.

It's quite possible to defunctionalize `TypeError` today. (Somewhat painful, due to [GHC Trac #12048](https://ghc.haskell.org/trac/ghc/ticket/12048)). But possible.) However, singling `ErrorMessage` proves to be a real pain the butt because of this constructor:

```haskell
  | Text Symbol
```

What should the `Text` constructor demote to? In an ideal world, the field would magically demote to `Data.Text.Text`. But that's not how `singletons` currently works (the subject of #82).

We _could_ work around this issue by defining a separate `ErrorMessageTerm` datatype with a `TextTerm Data.Text.Text` constructor, but we'd lose something important—the ability to write code using a `typeError` function and have it work both at the value level and the type level. I'm putting particular emphasis on this since I think that is ultimately why @crockeea opened this ticket, so if we "fix" this issue but don't address that use case, I feel like we'd be missing the point.

**Comment by goldfirere** [comment]
This does sound like it would work, to me. I'm not sure about the usability trade offs (though something that works is better than nothing) -- others' opinions will have more weight.

**Comment by RyanGlScott** [comment]
@crockeea, as the person who opened this ticket, do you have an opinion on the design proposed in https://github.com/goldfirere/singletons/issues/160#issuecomment-338246486 ?

**Comment by crockeea** [comment]
I'm not sure what you mean by not being able to use `ShowError` at the value level, but your example basically does exactly what I was originally hoping for. Thanks for working on that; I think it will result in much more user-friendly errors in singletons code. 
Is the name conflict required? It'd be great to avoid, but I think it's a good proposal even with the name conflict.

**Comment by RyanGlScott** [comment]
> I'm not sure what you mean by not being able to use `ShowError` at the value level

Did you mean `ShowType`? It's because the definition of the `ShowType` constructor is this:

```haskell
data ErrorMessage s
  = ...
  | forall t. ShowType t
```

That is, its field is of a completely abstract type `t`, so if we're given a `ShowType x`, then we can't do anything useful with `x` (e.g., call `show` on it). Indeed, the only reason why this works at the type level is due to GHC magic, which we don't have access to in our world.

You might reasonably object "but why not just change it to something like `forall t. Show t => ShowType t`? That would be nice, except that GHC currently prevents us from doing this due to [Trac #13895](https://ghc.haskell.org/trac/ghc/ticket/13895).

For your use case, it actually wouldn't matter—`show x` would promote to `Show x`, which would work as you'd expect. And `typeError` at the value level would still be quite functional—you'd just print out `"<type>"` instead of pretty-printing a type name when you hit a `ShowType`. The only folks who'd be bitten by this are those who call `sTypeError` on a `ShowType`, but I suspect that will this be a minority of use cases.

> Is the name conflict required?

Yes. Unless you can think of a design which works around the issue :)

To recap https://github.com/goldfirere/singletons/issues/160#issuecomment-319579000, the thing that prevents you from using the original `ErrorMessage` definition is this constructor:

```haskell
  | Text Symbol
```

This can't be used at the value level, as there's simply no way to produce a value-level `Symbol`. Without `Text`, you can't really construct any meaningful `ErrorMessage`s, so this approach seems DOA.
